core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/StrictEvaluationStrategy.java

```diff
@@ -654,26 +555,151 @@ protected BindingSet convert(Statement st) {
-					try {
-						if (stIter3 != null) {
-							stIter3.close();
-						}
-					} finally {
-						try {
-							if (stIter2 != null) {
-								stIter2.close();
-							}
-						} finally {
-							if (stIter1 != null) {
-								stIter1.close();
-							}
-						}
+					if (stIter1 != null) {
+						stIter1.close();
 					}
 				}
 			}
 		}
 	}
 
+	/**
+	 * Generate a predicate that tests for Named contexts are matched by retrieving all statements from the store and
+	 * filtering out the statements that do not have a context. Or. The same variable might have been used multiple
+	 * times in this StatementPattern, verify value equality in those cases.
+	 */
+	protected Predicate<Statement> filterOnStatements(StatementPattern statementPattern, final Var subjVar,
+			final Var predVar,
+			final Var objVar, final Var conVar, final Value subjValue, final Value predValue, final Value objValue,
+			Resource[] contexts) {
+		Predicate<Statement> filter = null;
+		if (contexts.length == 0 && statementPattern.getScope() == Scope.NAMED_CONTEXTS) {
+			filter = (st) -> st.getContext() != null;
+		}
+		filter = filterSameVariable(subjVar, predVar, objVar, conVar, subjValue, predValue, objValue, filter);
+		return filter;
+	}
+
+	protected Predicate<Statement> filterSameVariable(final Var subjVar, final Var predVar, final Var objVar,
+			final Var conVar, final Value subjValue, final Value predValue, final Value objValue,
+			Predicate<Statement> filter) {
+		if (subjVar != null && (subjVar.equals(predVar) || subjVar.equals(objVar) || subjVar.equals(conVar))
+				&& subjValue == null) {
+			filter = andThen(filter, (st) -> {
+				Resource subj = st.getSubject();
+				IRI pred = st.getPredicate();
+				Value obj = st.getObject();
+				Resource context = st.getContext();
+
+				if (subjVar.equals(predVar) && !subj.equals(pred)) {
+					return false;
+				}
+				if (subjVar.equals(objVar) && !subj.equals(obj)) {
+					return false;
+				}
+				if (subjVar.equals(conVar) && !subj.equals(context)) {
+					return false;
+				}
+				return true;
+			});
+		}
+		if (predVar != null && (predVar.equals(objVar) || predVar.equals(conVar)) && predValue == null) {
+			filter = andThen(filter, (st) -> {
+				IRI pred = st.getPredicate();
+				Value obj = st.getObject();
+				Resource context = st.getContext();
+
+				if (predVar.equals(objVar) && !pred.equals(obj)) {
+					return false;
+				}
+				if (predVar.equals(conVar) && !pred.equals(context)) {
+					return false;
+				}
+				return true;
+			});
+		}
+		if (objVar != null && (objVar.equals(conVar)) && objValue == null) {
+			filter = andThen(filter, (st) -> {
+				Value obj = st.getObject();
+				Resource context = st.getContext();
+				if (objVar.equals(conVar) && !obj.equals(context)) {
+					return false;
+				}
+				return true;
+			});
+		}
+		return filter;
+	}
+
+	private Predicate<Statement> andThen(Predicate<Statement> pred, Predicate<Statement> and) {
+		if (pred == null)
+			return and;
```

Comment:
Minor editorial: add braces.

Reply:
Sorry, work code style. Fixed an other case of this as well.